source/MRMesh/MRObjectGcode.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+#include "MRObjectLinesHolder.h"
+#include "MRGcodeExecutor.h"
+#include "MRColor.h"
+
+namespace MR
+{
+
+using GcodeSource = std::vector<std::string>;
+
+/// an object that stores a g-code
+/// \ingroup DataModelGroup
+class MRMESH_CLASS ObjectGcode : public ObjectLinesHolder
+{
+public:
+    MRMESH_API ObjectGcode();
+    ObjectGcode( ObjectGcode&& ) = default;
+    ObjectGcode& operator=( ObjectGcode&& ) = default;
+    virtual ~ObjectGcode() = default;
+
+    constexpr static const char* TypeName() noexcept { return "ObjectGcode"; }
+    virtual const char* typeName() const override { return TypeName(); }
+
+    MRMESH_API virtual std::shared_ptr<Object> clone() const override;
+    MRMESH_API virtual std::shared_ptr<Object> shallowClone() const override;
+
+    MRMESH_API virtual void setGcodeSource( const std::shared_ptr<GcodeSource>& gcodeSource );
+
+    virtual const std::shared_ptr<GcodeSource>& gcodeSource() { return gcodeSource_; }
+
+    MRMESH_API virtual void setDirtyFlags( uint32_t mask ) override;
+
+    /// \note this ctor is public only for std::make_shared used inside clone()
+    ObjectGcode( ProtectedStruct, const ObjectGcode& obj ) : ObjectGcode( obj ) {}
+
+    MRMESH_API virtual std::vector<std::string> getInfoLines() const override;
+    virtual std::string getClassName() const override { return "G-code"; }
+
+    // set drawing feedrate as gradient of brightness
+    MRMESH_API void setFeedrateGradient( bool feedrateGradient );
+    bool getFeedrateGradient() { return feedrateGradient_; }
+
+    MRMESH_API void setColor( Color color, bool work = true );
+    Color getColor( bool work = true ) { return work ? workColor_ : idleColor_; }
```

Comment:
please add comments about work argument, and `get` should be const

Reply:
also take color as const ref, and return color as const ref

Reply:
fixed